// tonfura-sdk/dist/cjs/TonfuraProvider-df3debf9.js
'use strict';

var axios = require('axios');
var index = require('./index-3549a2da.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var axios__default = /*#__PURE__*/_interopDefaultLegacy(axios);

/**
 * SDK's custom implementation of 'TonfuraProvider'.
 *
 * Do not call this constructor directly. Instead, instantiate an instance of
 * {@link Tonfura} and call {@link Tonfura.config.getProvider()}.
 *
 * @public
 */
class TonfuraProvider {
    /** @internal */
    constructor(config) {
        const apiKey = config.apiKey;
        const network = config.network;
        this.baseUrl = index.getTonfuraHttpUrl(network, apiKey);
        if (config.url !== undefined) {
            this.baseUrl = config.url;
        }
        this.maxRetries = config.maxRetries;
    }
    /** @internal */
    sendJsonRpcRequest(method, params) {
        return axios__default["default"].post(this.baseUrl, {
            jsonrpc: '2.0',
            id: 0,
            method,
            params
        });
    }
}

exports.TonfuraProvider = TonfuraProvider;
